### PR TITLE
JAMES-2159 Fix contacts

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java [hunk 1]
```diff
@@ -93,13 +93,24 @@ public void service(Mail mail) throws MessagingException {
...
                 .filter(Throwing.predicate(sender -> hasRecipients(mimeMessage)))
                 .map(Throwing.function(sender -> new ExtractedContacts(sender, recipients(mimeMessage))))
                 .map(Throwing.function(message -> objectMapper.writeValueAsString(message)));
     }
 
+    private Optional<Address> getSender(MimeMessage mimeMessage) throws MessagingException {
+        Optional<Address> sender = Optional.ofNullable(mimeMessage.getSender());
+        if (sender.isPresent()) {
+            return sender;
+        } else {
```
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java]: else not needed

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java [hunk 2]
```diff
@@ -93,13 +93,24 @@ public void service(Mail mail) throws MessagingException {
...
+            return Optional.ofNullable(mimeMessage.getFrom())
```
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java]: extract to _getFrom_

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java [hunk 3]
```diff
@@ -183,6 +183,24 @@ public void serviceShouldSupportMultipleFromField() throws Exception {
...
+            .build();
+        FakeMail mail = FakeMail.builder().mimeMessage(message)
+            .sender(new MailAddress(SENDER))
+            .recipient(new MailAddress(TO))
+            .build();
+        mailet.init(mailetConfig);
+
+        mailet.service(mail);
+
+        assertThatJson(mail.getAttribute(ATTRIBUTE)).isEqualTo(null);
```
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java]: The value of the attribute is null or the attribute is not present?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java [hunk 4]
```diff
@@ -215,6 +215,23 @@ public void serviceShouldNotAddTheAttributeWhenNoRecipient() throws Exception {
...
+            .build();
+        FakeMail mail = FakeMail.builder().mimeMessage(message)
+            .sender(new MailAddress(SENDER))
+            .recipient(new MailAddress(TO))
+            .build();
+        mailet.init(mailetConfig);
+
+        mailet.service(mail);
 
+        assertThatJson(mail.getAttribute(ATTRIBUTE)).isEqualTo(null);
```
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java]: same

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java [hunk 5]
```diff
@@ -93,13 +93,24 @@ public void service(Mail mail) throws MessagingException {
...
+                .filter(addresses -> addresses.length > 0)
+                .map(addresses -> addresses[0]);
+        }
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java]: so if I send a message without a sender and with a forged From, I can add contacts to an other user?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java (hunk 4 above)
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java]: assertThat isNull

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java [hunk 6]
```diff
@@ -92,33 +101,68 @@ public void service(Mail mail) throws MessagingException {
...
+                .flatMap(this::toMailboxStream)
+                .findFirst();
         }
     }
 
-    private boolean hasRecipients(MimeMessage mimeMessage) throws MessagingException {
-        return mimeMessage.getAllRecipients().length > 0;
+    private boolean hasRecipients(Message message) throws MessagingException {
+        return !message.getTo().isEmpty()
+            ||!message.getCc().isEmpty()
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java]: space after ||

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java [hunk 7]
```diff
@@ -92,33 +101,68 @@ public void service(Mail mail) throws MessagingException {
...
+            ||!message.getBcc().isEmpty();
     }
 
-    private ImmutableList<String> recipients(MimeMessage mimeMessage) throws MessagingException {
-        return Arrays.stream(mimeMessage.getAllRecipients())
-                .map(Address::toString)
-                .collect(Guavate.toImmutableList());
+    private ImmutableList<String> recipients(Message message) throws MessagingException {
+        return Stream.concat(toStream(message.getTo()),
+            Stream.concat(toStream(message.getCc()),
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java]: indent

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java [hunk 8]
```diff
@@ -162,6 +162,26 @@ public void serviceShouldUseFromFieldWhenNoSender() throws Exception {
...
+            .sender(new MailAddress(SENDER))
+            .recipient(new MailAddress(TO))
+            .build();
+        mailet.init(mailetConfig);
+
+        String expectedMessage = "{\"userEmail\" : \"" + SENDER + "\", \"emails\" : [ \"To <" + TO + ">\" ]}";
+        mailet.service(mail);
+
+        assertThatJson(mail.getAttribute(ATTRIBUTE).toString()).isEqualTo(expectedMessage);
+    }
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java]: are you able to add a test with a group address type?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java (hunk 3 above)
Comment (chibenwa) [reply mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java]: Null means not present.

Here there is also some concerns with assertj on serializable stull.

Maybe we need to assert attributes names does not contain Blah...

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java (hunk 5 above)
Comment (chibenwa) [reply mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java]: Ah... That's not stupid.

Maybe we need **TO GET SENDER FROM THE ENVELOPE** but recipients from the Mime Message

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java (hunk 4 above)
Comment (chibenwa) [reply mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java]: Well, see @aduprat  comment answer

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java (hunk 8 above)
Comment (chibenwa) [reply mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java]: I don't know the damn syntax

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java (hunk 3 above)
Comment (chibenwa) [reply mailet/standard/src/test/java/org/apache/james/transport/mailets/ContactExtractorTest.java]: Oh crap. That's why I thought assertj was limited with serializable....

Face palm

**Pull request closed without merging**